Fix for Issue 2156

kaby76 opened the pull request:
This is a fix for #2156, a change to the apex/, java/java/, java/java8/, and java/java9/ grammars so they may be target independent.

* The grammars now use the 32-bit Unicode code point and general categories. This replaces the older [predicate action code](https://github.com/antlr/grammars-v4/blob/9e449d012348fdb143a3375bed73ee01ba562196/java/java9/Java9Lexer.g4#L487) to test whether the code point is valid.
* The character sets corresponding to [Character.isJavaIdentifierStart()](https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isJavaIdentifierStart(int)) and [Character.isJavaIdentifierPart()](https://docs.oracle.com/javase/7/docs/api/java/lang/Character.html#isJavaIdentifierPart(int)), which are methods actually referenced in the Java Language Spec, were experimentally derived using the programs in grammars-v4/java/support/. The sets were used to write the rules for JavaLetter and JavaLetterOrDigit in each of the four grammars. Note, the Java program that was in java/java8/ to compute the sets for these methods was moved to java/support/.
* The new rules in the optimized java/java/ grammar resulted in no regressions and no performance degradation when tested on the OpenJDK source version 16+. Still, about 10% of the OpenJDK does not parse with the java/java/ grammar.
* The Antlr4cs target (Harwell's version of the Antlr4 tool and runtime for C#) does not support 32-bit Unicode grammar syntax. The old grammars were placed in the Antlr4cs/ directory with support code, and adjusted so they would work for Antlr4cs (splitting the grammar, adding base class, etc).
* Note, in splitting the apex/ grammar, there was a problem with '<>' not being defined in the lexer. I added a rule to define a token type for the string literal.
* The workflow was updated to use the latest version of trgen. Harwell's tool is not being actively updated, so I modified the trgen tool to avoid collisions of combined/splitted grammars, e.g., in apex/.
* There are no tests for the various encodings in the CI build. But, I have tested the grammars with utf-8, -16, -32 encodings using CSharp and Java, and the templates in _scripts/templates for those two targets have been updated to recognize a "-encoding" option. I could not get encodings to work with Dart, and I have not written code for recognizing encodings for the other targets. Antlr4cs does not support encodings.
* Additional work must be done to the Java grammar: the language is now on version 16, and a large number of files in the OpenJDK source do not parse. java/java8/ and java/java9/ are not optimized grammars. The best way to proceed on updating the Java grammars is to scrape the grammar from the Java Language Spec's, then apply refactorings to optimize in an automated fashion.
* This PR took a long time to develop. A big problem that thwarted my effort is described in issue #2182. (I have a fix to be added soon.)
* Regarding the use of a fixed character set rule for JavaLetter and JavaLetterOrDigit as opposed to following faithfully the JLS. I have written the JLS committee to ask why they continue to define those rules operationally in the Spec. It seems plausible that they could change the spec at some point going forward to not define the language operationally as it makes it more difficult for language implementers to do the right thing.

studentmain commented on apex/apexLexer.g4:
```diff
@@ -0,0 +1,497 @@
+    ;
+
+fragment
+JavaLetter
```
https://github.com/antlr/grammars-v4/blob/e53d7a1228505bfc80d8637808ef60e7eea92cc2/java/java8/Java8Lexer.g4#L469

We can make it target independent.

kaby76 replied:
Goodness. This code--and Java9--both need to be cleaned up, pronto! In addition, the commented out code in Java8Lexer.g4 needs to be just removed. At first, I was going to ask you how this helps because I was seeing the semantic action code above the [IdentifierStart](https://github.com/antlr/grammars-v4/blob/e53d7a1228505bfc80d8637808ef60e7eea92cc2/java/java8/Java8Lexer.g4#L469) rule. But, then I noticed it was commented out. The commented-out code should all be removed. Really, let's use the power of version control if one needs to see an older version of the grammar.

I'll change this issue and broaden it to include Java8, Java9, and Apex.

kaby76 commented:
I'm now going through the different versions and decisions of Identifiers in the various Java grammars.

According to the spec, [a JavaLetter is defined as](https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.8):

    A "Java letter" is a character for which the method Character.isJavaIdentifierStart(int) returns true.

Below that, however, the spec then goes into the details of how to actually implement the method:

    The "Java letters" include uppercase and lowercase ASCII Latin letters A-Z (\u0041-\u005a), and a-z (\u0061-\u007a), and, for historical reasons, the ASCII underscore (_, or \u005f) and dollar sign ($, or \u0024). The $ sign should be used only in mechanically generated source code or, rarely, to access pre-existing names on legacy systems.
    The "Java digits" include the ASCII digits 0-9 (\u0030-\u0039).
    Letters and digits may be drawn from the entire Unicode character set, which supports most writing scripts in use in the world today, including the large sets for Chinese, Japanese, and Korean. This allows programmers to use identifiers in their programs that are written in their native languages.
    An identifier cannot have the same spelling (Unicode character sequence) as a keyword (§3.9), boolean literal (§3.10.3), or the null literal (§3.10.7), or a compile-time error occurs.
    Two identifiers are the same only if they are identical, that is, have the same Unicode character for each letter or digit. Identifiers that have the same external appearance may yet be different.

It's seems fair to choose an implementation either using a semantic predicate or a predicate-free implementation.

There are two predicate-free implementations for JavaLetter/Letter, one in [java/java/](https://github.com/antlr/grammars-v4/blob/master/java/java/JavaLexer.g4#L207) and the other in [java/java8](https://github.com/antlr/grammars-v4/blob/master/java/java8/Java8Lexer.g4#L469), and they are different. It's not clear why.

In both predicate-free implementations, they do not appear to address UTF-32. UTF-16 is a variable-length encoding, which UTF-32 is a fixed length encoding. I thought that Antlr worked with UTF-32 in order to jump to the N-th character directly, but I will need to check.

studentmain commented:
JavaLetter in java8 are generated by calling those methods. The detail mentioned in the spec is incomplete, e.g. `int 蛤=1926;` is valid, `int 蛤。=1926;` is invalid. We'd better reading the source code to see how they actually implement it.

About UTF-32 (in fact it's about supplementary plane, not exactly about encoding), seems ANTLR can't even represent them, we need unicode literal which can represent code point bigger than \uffff first.

kaby76 commented:
> We'd better reading the source code to see how they actually implement it.

I agree, it's best to really see how Java works. Going through that now, e.g. making notes how Javac handles different input encodings. This is going to take a little while.


> About UTF-32 (in fact it's about supplementary plane, not exactly about encoding), seems ANTLR can't even represent them, we need unicode literal which can represent code point bigger than \uffff first.

kaby76 commented on apex/CSharp/ApexLexerBase.cs:
```diff
@@ -0,0 +1,100 @@
+
+    private class Character
+    {
+        public static bool isJavaIdentifierPart(int c)
+        {
+            if (Char.IsLetter((char)c))
+                return true;
+            else if (c == (int)'$')
+                return true;
+            else if (c == (int)'_')
+                return true;
+            else if (Char.IsDigit((char)c))
+                return true;
+            else if (Char.IsNumber((char)c))
+                return true;
+            return false;
+        }
+
+        public static bool isJavaIdentifierStart(int c)
+        {
+            if (Char.IsLetter((char)c))
+                return true;
+            else if (c == (int)'$')
+                return true;
+            else if (c == (int)'_')
+                return true;
+            return false;
+        }
```
Yes, better, but it does rely on the C# runtime which might have a different character classification from Java. So, it really goes back to what Java does. I'm looking at all combinations of javac -encoding on programs with BOM, no BOM, 8, 16, 32, and reading the OpenJDK source, starting with [isJavaIdentifierStart](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Character.java#L10070) => [of()](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/lang/Character.java#L10070) => [*.isJavaIdentifierStart](https://github.com/openjdk/jdk/blob/master/make/data/characterdata/CharacterData00.java.template#L97), etc. I will also try to compute the rule from the OpenJDK methods isJavaIdentifierStart() and isJavaIdentifierPart(). I do not understand why the spec presents two ways of defining `JavaLetter`. Terrible.

studentmain commented:
Their API reference looks much better... https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#isJavaIdentifierStart-int-

> A character may start a Java identifier if and only if one of the following conditions is true:
>
>   - isLetter(codePoint) returns true
>   - getType(codePoint) returns LETTER_NUMBER
>   - the referenced character is a currency symbol (such as '$')
>   - the referenced character is a connecting punctuation character (such as '_'). 
>
> These conditions are tested against the character information from version 6.2 of the Unicode Standard.

So isJavaIdentifierStart return true when character is in one of Unicode category L, Nl, Sc, Pc.

For isJavaIdentifierPart:

> - it is a letter
> - it is a currency symbol (such as '$')
> - it is a connecting punctuation character (such as '_')
> - it is a digit
> - it is a numeric letter (such as a Roman numeral character)
> - it is a combining mark
> - it is a non-spacing mark
> - isIdentifierIgnorable(codePoint) returns true for the code point

Corresponding Unicode categories are L, Sc, Pc, Nd, Nl, Mc, Mn. 

isIdentifierIgnorable's corresponding ranges are
\u0000-\u0008, \u000e-\u001b, \u007f-\u009f, and category Cf.

> - ISO control characters that are not whitespace
>   - '\u0000' through '\u0008'
>   - '\u000E' through '\u001B'
>   - '\u007F' through '\u009F'
> - all characters that have the FORMAT general category value

kaby76 commented:
@studentmain That looks good. I can _sort of_ see why the Java Spec is done this way: it's data-driven. But, the Java Spec should just use the damn Unicode general categories. Looking at the old mail archive groups for OpenJDK on using G.C.s and history of all this stuff, but there's no search in the [website](http://mail.openjdk.java.net/mailman/listinfo), and it's in a bunch of chunks, so I can't download one archive for a search. Geez.

kaby76 commented:
I wrote a program to experimentally derive the character classes for isJavaIdentifierStart(). The result is Ll | Lm | Lo | Lt | Lu | Nl | Pc | Sc.

For isJavaIdentifierPart, it is Cc | Cf | Ll | Lm | Lo | Lt | Lu | Mc | Mn | Nd | Nl | Pc | Sc

	import java.util.*;

	class HelloWorld {
		public static void main(String[] args) {
			int ctr = 0;

			Map<Integer, String> unicodeCategories = new HashMap<>();
			unicodeCategories.put((int)Character.UNASSIGNED, "Cn"); // 0
			unicodeCategories.put((int)Character.UPPERCASE_LETTER, "Lu"); // 1
			unicodeCategories.put((int)Character.LOWERCASE_LETTER, "Ll"); // 2
			unicodeCategories.put((int)Character.TITLECASE_LETTER, "Lt"); // 3
			unicodeCategories.put((int)Character.MODIFIER_LETTER, "Lm"); // 4
			unicodeCategories.put((int)Character.OTHER_LETTER, "Lo"); // 5
			unicodeCategories.put((int)Character.NON_SPACING_MARK, "Mn"); // 6
			unicodeCategories.put((int)Character.ENCLOSING_MARK, "Me"); // 7
			unicodeCategories.put((int)Character.COMBINING_SPACING_MARK, "Mc"); // 8
			unicodeCategories.put((int)Character.DECIMAL_DIGIT_NUMBER, "Nd"); // 9
			unicodeCategories.put((int)Character.LETTER_NUMBER, "Nl"); // 10
			unicodeCategories.put((int)Character.OTHER_NUMBER, "No"); // 11
			unicodeCategories.put((int)Character.SPACE_SEPARATOR, "Zs"); // 12
			unicodeCategories.put((int)Character.LINE_SEPARATOR, "Zl"); // 13
			unicodeCategories.put((int)Character.PARAGRAPH_SEPARATOR, "Zp"); // 14
			unicodeCategories.put((int)Character.CONTROL, "Cc"); // 15
			unicodeCategories.put((int)Character.FORMAT, "Cf"); // 16
			unicodeCategories.put((int)Character.PRIVATE_USE, "Co"); // 18
			unicodeCategories.put((int)Character.SURROGATE, "Cs"); // 19
			unicodeCategories.put((int)Character.DASH_PUNCTUATION, "Pd"); // 20
			unicodeCategories.put((int)Character.START_PUNCTUATION, "Ps"); // 21
			unicodeCategories.put((int)Character.END_PUNCTUATION, "Pe"); // 22
			unicodeCategories.put((int)Character.CONNECTOR_PUNCTUATION, "Pc"); // 23
			unicodeCategories.put((int)Character.OTHER_PUNCTUATION, "Po"); // 24
			unicodeCategories.put((int)Character.MATH_SYMBOL, "Sm"); // 25
			unicodeCategories.put((int)Character.CURRENCY_SYMBOL, "Sc"); // 26
			unicodeCategories.put((int)Character.MODIFIER_SYMBOL, "Sk"); // 27
			unicodeCategories.put((int)Character.OTHER_SYMBOL, "So"); // 28
			unicodeCategories.put((int)Character.INITIAL_QUOTE_PUNCTUATION, "Pi"); // 29
			unicodeCategories.put((int)Character.FINAL_QUOTE_PUNCTUATION, "Pf"); // 30

			for (int codePoint = 0; codePoint < 0x10FFFF; codePoint++)
			{
				if (Character.isJavaIdentifierStart(codePoint))
				{
					System.out.print(String.format("0x%08X", codePoint));
					var general_class = Character.getType(codePoint);
					var gc_name = unicodeCategories.get(general_class);
					if (gc_name == null) gc_name = String.format("0x%02X", general_class);
					System.out.println(" " + gc_name);
					ctr++;
				}
			}
			System.out.println("Number of UC codepoints " + ctr);
		}
	}

studentmain commented:
> I wrote a program to experimentally derive the character classes for isJavaIdentifierStart().

Check if all character in those categories are JavaIdentifierStart, here's a counter example: \u000a, \n, it's in Cc, obviously not a JavaIdentifierPart.

kaby76 commented:
Right, they're not equivalent, so it will need to be narrowed with individual code points rather than the general category. Almost there. I will complain to someone on the spec committee about this operational def. I just don't get it.

In any case, Antlr CharStream is char[], char = 2 bytes in Java and C#. So, we'll need to use a different stream to handle >2 byte codepoints, or use CharStream and work with variable length values. Otherwise, how are we going to parse UTF-32 files?

studentmain commented:
did a quick test, isJavaIdentifierStart are matched with those categories, isJavaIdentifierPart has these exception, same as the document said.
```
\u0009
\u000A
\u000B
\u000C
\u000D
\u001C
\u001D
\u001E
\u001F
```

studentmain commented:
> Otherwise, how are we going to parse UTF-32 files

https://github.com/antlr/antlr4/pull/1633

kaby76 commented:
I've checked the code points to verify whole sets and generate a list of exceptions. For isJavaIdentifierStart(), I get Lu | Ll | Lt | Lm | Lo | Nl | Pc | Sc.

For isJavaIdentifierPart(), I get:

	Lu
	Ll
	Lt
	Lm
	Lo
	Mn
	Mc
	Nd
	Nl
	Cf
	Pc
	Sc
	Plus 0x00000000
	Plus 0x00000001
	Plus 0x00000002
	Plus 0x00000003
	Plus 0x00000004
	Plus 0x00000005
	Plus 0x00000006
	Plus 0x00000007
	Plus 0x00000008
	Plus 0x0000000E
	Plus 0x0000000F
	Plus 0x00000010
	Plus 0x00000011
	Plus 0x00000012
	Plus 0x00000013
	Plus 0x00000014
	Plus 0x00000015
	Plus 0x00000016
	Plus 0x00000017
	Plus 0x00000018
	Plus 0x00000019
	Plus 0x0000001A
	Plus 0x0000001B
	Plus 0x0000007F
	Plus 0x00000080
	Plus 0x00000081
	Plus 0x00000082
	Plus 0x00000083
	Plus 0x00000084
	Plus 0x00000085
	Plus 0x00000086
	Plus 0x00000087
	Plus 0x00000088
	Plus 0x00000089
	Plus 0x0000008A
	Plus 0x0000008B
	Plus 0x0000008C
	Plus 0x0000008D
	Plus 0x0000008E
	Plus 0x0000008F
	Plus 0x00000090
	Plus 0x00000091
	Plus 0x00000092
	Plus 0x00000093
	Plus 0x00000094
	Plus 0x00000095
	Plus 0x00000096
	Plus 0x00000097
	Plus 0x00000098
	Plus 0x00000099
	Plus 0x0000009A
	Plus 0x0000009B
	Plus 0x0000009C
	Plus 0x0000009D
	Plus 0x0000009E
	Plus 0x0000009F

So, glad I checked. I just don't trust any documentation--it's never up to date.

	import java.util.*;

	class HelloWorld {
		public static void main(String[] args) {
			int ctr = 0;

			Map<Integer, String> unicodeCategories = new HashMap<>();
			unicodeCategories.put((int)Character.UNASSIGNED, "Cn"); // 0
			unicodeCategories.put((int)Character.UPPERCASE_LETTER, "Lu"); // 1
			unicodeCategories.put((int)Character.LOWERCASE_LETTER, "Ll"); // 2
			unicodeCategories.put((int)Character.TITLECASE_LETTER, "Lt"); // 3
			unicodeCategories.put((int)Character.MODIFIER_LETTER, "Lm"); // 4
			unicodeCategories.put((int)Character.OTHER_LETTER, "Lo"); // 5
			unicodeCategories.put((int)Character.NON_SPACING_MARK, "Mn"); // 6
			unicodeCategories.put((int)Character.ENCLOSING_MARK, "Me"); // 7
			unicodeCategories.put((int)Character.COMBINING_SPACING_MARK, "Mc"); // 8
			unicodeCategories.put((int)Character.DECIMAL_DIGIT_NUMBER, "Nd"); // 9
			unicodeCategories.put((int)Character.LETTER_NUMBER, "Nl"); // 10
			unicodeCategories.put((int)Character.OTHER_NUMBER, "No"); // 11
			unicodeCategories.put((int)Character.SPACE_SEPARATOR, "Zs"); // 12
			unicodeCategories.put((int)Character.LINE_SEPARATOR, "Zl"); // 13
			unicodeCategories.put((int)Character.PARAGRAPH_SEPARATOR, "Zp"); // 14
			unicodeCategories.put((int)Character.CONTROL, "Cc"); // 15
			unicodeCategories.put((int)Character.FORMAT, "Cf"); // 16
			unicodeCategories.put((int)Character.PRIVATE_USE, "Co"); // 18
			unicodeCategories.put((int)Character.SURROGATE, "Cs"); // 19
			unicodeCategories.put((int)Character.DASH_PUNCTUATION, "Pd"); // 20
			unicodeCategories.put((int)Character.START_PUNCTUATION, "Ps"); // 21
			unicodeCategories.put((int)Character.END_PUNCTUATION, "Pe"); // 22
			unicodeCategories.put((int)Character.CONNECTOR_PUNCTUATION, "Pc"); // 23
			unicodeCategories.put((int)Character.OTHER_PUNCTUATION, "Po"); // 24
			unicodeCategories.put((int)Character.MATH_SYMBOL, "Sm"); // 25
			unicodeCategories.put((int)Character.CURRENCY_SYMBOL, "Sc"); // 26
			unicodeCategories.put((int)Character.MODIFIER_SYMBOL, "Sk"); // 27
			unicodeCategories.put((int)Character.OTHER_SYMBOL, "So"); // 28
			unicodeCategories.put((int)Character.INITIAL_QUOTE_PUNCTUATION, "Pi"); // 29
			unicodeCategories.put((int)Character.FINAL_QUOTE_PUNCTUATION, "Pf"); // 30

			Set<Integer> full_category = new HashSet<Integer>();

			for (int category = 0; category < 31; ++category)
			{
				// Verify every code point in category is also in isJavaIdentiferStart().
				boolean all = true;
				boolean any = false;
				for (int codePoint = 0; codePoint < 0x10FFFF; codePoint++)
				{
					var general_class = Character.getType(codePoint);
					if (category == general_class)
					{
						any = true;
						if (! Character.isJavaIdentifierPart(codePoint))
						{
							all = false;
							break;
						}
					}
				}
				if (! all)
				{
					continue;
				}
				if (! any)
				{
					continue;
				}
				var name = unicodeCategories.get(category);
				if (name == null)
				{
					continue;
				}
				System.out.println(name);
				full_category.add(category);
			}

			// Now go through all code points and verify that it appears in one full category.
			// Otherwise, we need to now add that to the special case list.
			for (int codePoint = 0; codePoint < 0x10FFFF; codePoint++)
			{
				if (! Character.isJavaIdentifierPart(codePoint))
				{
					continue;
				}
				var general_class = Character.getType(codePoint);
				if (full_category.contains(general_class))
				{
					continue;
				}
				System.out.println(String.format("Plus 0x%08X", codePoint));
			}

			//System.out.println("Number of UC codepoints " + ctr);
		}
	}

kaby76 commented:
> [antlr/antlr4#1633](https://github.com/antlr/antlr4/pull/1633)

Excellent. I'll give it a whirl.

kaby76 commented:
Using grammars-v4/java/java/, modified it with:

    fragment JavaLetter : [\p{Lu}] | [\p{Ll}] | [\p{Lt}] | [\p{Lm}] | [\p{Lo}] | [\p{Nl}] | [\p{Pc}] | [\p{Sc}] ;
    fragment JavaLetterOrDigit : [\p{Lu}] | [\p{Ll}] | [\p{Lt}] | [\p{Lm}] | [\p{Lo}] | [\p{Mn}] | [\p{Mc}] | [\p{Nd}] | [\p{Nl}] | [\p{Cf}] | [\p{Pc}] | [\p{Sc}] | [\u{00000000}-\u{00000008}] | [\u{0000000E}-\u{0000001B}] | [\u{0000007F}-\u{0000009F}] ;

In a C# driver,

    str = CharStreams.fromPath(file_name, Encoding....);

A sanity parse of "HelloWorld.java" using UTF8, 16, 32 encodings works. Need to check range sets in JavaLetterOrDigit fully.

kaby76 commented:
I updated the grammars in apex, java/{java,java8,java9}/ with predicate-free versions we discussed. If the grammars were split before this PR, I left it that way, however, they could all be combined (e.g., Java9Lexer.g4 + Java9Parser.g4 => Java9.g4). I moved the code point checking software from java/java8/ to java/support/ since this software is java-version independent. I updated the _scripts/templates/CSharp/Program.cs to include an option "-encoding" to test the parser against UTF-8, -16, -32 file encodings, and I will need to update the other targets with the option. But, first, I need to check the grammars against my 50k Java source file test suite.

kaby76 commented:
I decided to first start testing the java/java/ grammar against the OpenJDK, which in itself has 50k files. I've been running a test of the C# parser using the before PR/after PR grammars on the JDK source, and it's been taking all day. It's half done, but from I can see, the parsers fails identically, and the parse times appear similar. I haven't calculated a T-test yet, but I will once the test is done.

studentmain commented on _scripts/templates/CSharp/Program.cs:
```diff
@@ -50,6 +51,17 @@ static void Main(string[] args)
+            else if (args[i].Equals("-encoding"))
+            {
+                ++i;
+                if (args[i].ToLower().Equals("utf-8"))
```
https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.getencoding?view=net-5.0#System_Text_Encoding_GetEncoding_System_String_

kaby76 commented:
Java and CSharp templates now have an "encoding" parameter for testing, although it's only for when the "-file" command-line arg is specified. The code in the templates look up the IANA name using the appropriate runtime methods. Antlr4cs does not support encodings. Further, the Antlr4cs tool does not recognize the Unicode code point notations in the grammar. So, the grammars have to forked for this target. It would seem Dart and the Dart Antlr runtime would support encodings, but it turns out it doesn't work. The method Encoding.getByName(), should return an encoding that I could pass to the Antlr input stream creation method, but getByName() returns null for every encoding name, e.g., "utf-32"--except "utf-8", which does work. That sucks.

kaby76 commented:
stringtemplate/ needs several fixes:

* Grammar ST{Parser,Lexer}.g4 is a grammar for regular StringTemplate.
* Grammar STG{Parser,Lexer}.g4 is a grammar for StringTemplate group.
* These should be separated into two sub-directories, e.g., follow antlr/antlr{2,3,4}.
* Only ST{Parser,Lexer}.g4 is actually tested on input, and __only one test file is provided__.
* I don't know what version these grammars are for. It might be best to start from scratch, convert the Antlr3 grammars of StringTemplate using trconvert, and work with that.
* There's no readme.
* Lexunicode.g4 (an include file) is included in LexBasic.g4, but *not one* rule is used!
* LexBasic.g4 uses a couple of predicates to check Java letter, but it differs from Java/java with JavaLetterOrDigit.
* It places some of the predicates in a lexer base class.

The two grammars here should be updated in their own issue.

I won't be updating stringtemplate/ with this PR.

There seems to be a problem with the CI build with JavaScript. I noticed this before--some JavaScript tests do not terminate, or just that they take super long. Normally I would kill the test and change the scripts to avoid the problem. But, Google Actions does not have a `kill -9`. This is really dumb. And, Antlr JavaScript is a mess.

studentmain commented:
> I won't be updating stringtemplate/ with this PR.

OK, let's open another PR for it.

> not have a kill -9

It has a cancel button, but only for admin.

kaby76 commented:
> > not have a kill -9
> 
> It has a cancel button, but only for admin.

It doesn't make any sense: I created the PR, but I have no way to control it, like Frankenstein's monster.

kaby76 commented:
It just occurred to me that to stop a runaway Github Action in the future, maybe I can just close the PR? I.e., __to kill Frankenstein's monster, blow up the castle__? I guess you have to think like a Github developer. To whoever killed the build, thank you kindly! Onwards.

The build failed in grammars-g4/agc/ JavaScript target? G.H.C! Renamed "register" to "register_" there. What happened?

I'm removing the CI test of the JavaScript target. It's just too unstable. Until the self-mutating ATN bug is fixed, and we figure out how to not repeat this again, we should avoid the JS target. 

-Ken

studentmain commented:
> maybe I can just close the PR?

It won't kill the build afaik. Adding new commit won't do so either...

About CI for JS target, it's better to remove them now (after this got merged, in another pr). It's meaningless to ensure grammar works in a buggy environment. Ensure the environment works with already tested grammar first.

kaby76 commented:
@studentmain I haven't made any further changes since the last commit that caused the hang in macosx/JavaScript. I wasn't sure what to do regarding the hang, because you suggested on making a separate PR to remove the macosx/JavaScript from .github/workflows/main.yml. That should be done before this code is merged, because the build here could hang again or hang for the next person who does a commit. So, I wasn't sure what to do next. In the meanwhile, I was going to work out a PR for the JS Antlr4 runtime to fix a problem there ([3135](https://github.com/antlr/antlr4/issues/3135) and which might also work for [3150](https://github.com/antlr/antlr4/issues/3150)). What I would prefer to do is disable the JS tests here with this commit, make sure the build works, then change this PR to "ready", then work on fixing the Antlr4 JS runtime, then come back here to work other things. Antlr4 JS won't be stable till the next release, which might not be for a while. Thoughts?

studentmain commented:
> I'll make a PR to disable the JS target now

Got it.

kaby76 commented:
Status:

* I will need to make exceptions to the grammars for at least Antlr4cs, Dart, and likely others.
* I updated [the description of the issue](https://github.com/antlr/grammars-v4/issues/2156#issue-866669052).
* I have asked the Java standards committee to explain why they use an operational definition for JavaLetter and JavaLetterOrDigit. We shall see.

kaby76 commented:
@studentmain Please peruse the changes (when the build finishes). Not perfect, but it should work in CI. It does a reasonable job of making the grammars target-independent given what we have to live with. I don't have tests for different encodings in CI: there's no code developed for handling encodings for Antlr4cs (we don't test), Dart (doesn't work), Go, Python3, JavaScript (we avoid right now), and Cpp (which we don't test at all). It does work for Java and CSharp, but again not tested in CI because there's no notion of skip lists for different encoding tests. I'm starting to tire of this PR, kind of like beating a dead horse.

kaby76 commented:
Ugh. Broken build--many fringe cases that trgen has to deal with.

studentmain commented:
> handling encodings

In my opinion, it should be tested at runtime's repo.


Every Go tests and asm/masm 's Java tests failed. Maybe due to recently refactor in Go template, and masm's package name.

kaby76 commented:
> Every Go tests and asm/masm 's Java tests failed. Maybe due to recently refactor in Go template, and masm's package name.

I had to change the trgen program: if apex.g4 is default grammar, but split in Antl4cs, the Antl4cs build rules get confused if all grammars exist in the same build directory. I added a test to not copy combined when the grammar is splitted, but the check wasn't working in different situations.

kaby76 commented:
* "emptyStatement" must be renamed globally because it causes a compilation error for the Go target. (In java/java8 and java/java9.)
* The Go target fails for java/java/ on java\java\examples\ManyStringsConcat.java with an out-of-memory condition. I am keeping java/{java,java8,java9} grammars on the skip list for Go. Apex works fine because it is just one test.
* The reason the build took slow long with the last run is because of the out-of-memory condition. This feeds back to issue #2114. The driver should place parsing tasks in a separate thread or process, killed if it goes over a timeout period. It's basically unacceptable to allow the parse to continue indefinitely.

kaby76 commented:
@teverett or @KvanTTT Please kill the CI 789 build. Go is not terminating on Macosx due to memory thrashing while parsing Java. After I get this PR finished--without the problematic Go target on the Java grammars--I will immediately address the issue of runaway tests by creating a tool that kills a child process tree when it consumes too much time or resources. Thanks. --Ken

kaby76 commented:
@teverett or @KvanTTT Sorry to request again, but please kill CI 792 build https://github.com/antlr/grammars-v4/actions/runs/813077171. I tried to rebase this PR, but there were many conflicts and I ended up undoing the absolutely essential changes to skip the tests of the Java grammars with Go. I am almost done with this PR. And, I will have a PR shortly for #2182 to finally watchdog the tests so this will never ever happen again.

teverett commented:
Done, and thanks

> Ken Domino ***@***.***>
> May 5, 2021 at 7:55 AM
>
> @teverett <https://github.com/teverett> or @KvanTTT 
> <https://github.com/KvanTTT> Sorry to request again, but please kill 
> CI 792 build 
> https://github.com/antlr/grammars-v4/actions/runs/813077171. I tried 
> to rebase this PR, but there were many conflicts and I ended up 
> undoing the absolutely essential changes to skip the tests of the Java 
> grammars with Go. I am almost done with this PR. And, I will have a PR 
> shortly for #2182 <https://github.com/antlr/grammars-v4/issues/2182> 
> to finally watchdog the tests so this will never ever happen again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/antlr/grammars-v4/pull/2157#issuecomment-832707431>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJLKZTDWJ5FKRAJHFNIHM3TMFE5RANCNFSM43PZ36WQ>.
>
> Ken Domino ***@***.***>
> April 24, 2021 at 12:18 AM
>
> *NB: This description is incomplete.*
>
> This is a fix for #2156 
> <https://github.com/antlr/grammars-v4/issues/2156>, splitting the 
> grammar and making it target-language independent.
>
> ------------------------------------------------------------------------
>
>
>         You can view, comment on, or merge this pull request online at:
>
> https://github.com/antlr/grammars-v4/pull/2157
>
>
>         Commit Summary
>
>   * Fix for #2156.
>
>
>         File Changes
>
>   * *A* apex/CSharp/ApexLexerBase.cs
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-c942add3c8e904d18c2b7558b999bd1fa786d3194815f8efdce2a99b302ca3f6>
>     (100)
>   * *A* apex/Cpp/ApexLexerBase.cpp
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-175cc897af148499302654f41130ec23b7a143e12c87fef1477119879d25ed04>
>     (91)
>   * *A* apex/Cpp/ApexLexerBase.h
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-4896b86a24daecff9b006962fdbe439be278d0c070f2117b2d44b16057c864aa>
>     (24)
>   * *A* apex/Cpp/apexLexer.g4
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-8b03ad964307e4b2610619ecb51904812339ac9ccd85fc6138ebae32f276477b>
>     (502)
>   * *A* apex/Dart/ApexLexerBase.dart
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-b0b618f61beb6d8434bdbc0fe2c3f6c2e172d8f97d502c76852c7f01fba3323f>
>     (102)
>   * *A* apex/Dart/apexLexer.g4
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-6520665bf8d7a439f48c86226749769cc2c92e82806dd60dc928fd281af32699>
>     (501)
>   * *A* apex/Java/ApexLexerBase.java
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-8292259c4bf72109ab489689b84446097bf30bba51ca1b355e258ee26dbe120e>
>     (61)
>   * *A* apex/apexLexer.g4
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-e0b909223b13e7f8ef62a8f3f651039e8401afcb5f62ad2c3bd01909edb014ed>
>     (497)
>   * *R* apex/apexParser.g4
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-9195339ddbcf6c26d6661f10caf1b59f3c7e51609652b7f53505a2a0c3b79721>
>     (496)
>   * *M* apex/pom.xml
>     <https://github.com/antlr/grammars-v4/pull/2157/files#diff-da0159d3d3010527ba40ed199e53711fb8430230b15b3d6621c62c2d8f7fe472>
>     (3)
>
>
>         Patch Links:
>
>   * https://github.com/antlr/grammars-v4/pull/2157.patch
>   * https://github.com/antlr/grammars-v4/pull/2157.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/antlr/grammars-v4/pull/2157>, or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAJLKZTNU7SIECU3HG2SZHLTKJPBZANCNFSM43PZ36WQ>.
>

-- 
Tom Everett
***@***.*** ***@***.***>
<https://www.postbox-inc.com>

KvanTTT reviewed (changes requested):
It seems like there are erroneous commits in this pull request and a lot of changes. Could you please rebase this branch and leave only relevant changes?

kaby76 commented:
Most of these changes are from doing a rebase--I do not understand why git rebase has all this "git rebase master".

I will close this PR/Issue out completely and start completely from complete scratch, work with one grammar one issue at a time.

KvanTTT commented:
It's possible to remove excess commits using [interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History). But I prefer using GUI for doing such operation (JetBrains-like IDE or GitExtensions).
